io/StacReader.cpp

```diff
@@ -127,112 +123,206 @@ void StacReader::addArgs(ProgramArgs& args)
         "https://schemas.stacspec.org/v1.0.0/catalog-spec/json-schema/catalog.json");
+    args.add("collection_schema_url", "URL of collection schema you'd like to use for"
+        " JSON schema validation.", m_args->schemaUrls.collection,
+        "https://schemas.stacspec.org/v1.0.0/collection-spec/json-schema/collection.json");
     args.add("feature_schema_url", "URL of feature schema you'd like to use for"
-        " JSON schema validation.", m_args->featureSchemaUrl,
+        " JSON schema validation.", m_args->schemaUrls.item,
         "https://schemas.stacspec.org/v1.0.0/item-spec/json-schema/item.json");
-    args.add("requests", "Number of threads for fetching JSON files, Default: 8",
-        m_args->threads, 8);
-    args.addSynonym("requests", "threads");
 
+}
+
+
+void StacReader::addItem(stac::Item& item)
+{
+    std::string driver = item.driver();
+
+    Stage *stage = m_factory.createStage(driver);
+
+    if (!stage)
+    {
+        std::stringstream msg;
+        msg << "Unable to create driver '" << driver << "' for "
+            << "for asset located at '" << item.assetPath() <<"'";
+        throw pdal_error(msg.str());
+    }
+    Reader* reader = dynamic_cast<Reader*>(stage);
+    if (!reader)
+    {
+        std::stringstream msg;
+        msg << "Unable to cast stage of type '" << driver << "' to a reader";
+        throw pdal_error(msg.str());
+    }
+    reader->setOptions(item.options());
+
+    std::lock_guard<std::mutex> lock(m_p->m_mutex);
+    m_p->m_idList.push_back(item.id());
+    reader->setLog(log());
+    m_merge.setInput(*reader);
+    m_p->m_readerList.push_back(reader);
+}
 
+void StacReader::handleItem(NL::json stacJson, std::string itemPath)
+{
+    stac::Item item(stacJson, m_filename, *m_p->m_connector,
+        m_args->validateSchema);
+    if (item.init(*m_p->m_itemFilters, m_args->rawReaderArgs, m_args->schemaUrls))
+        addItem(item);
 }
 
 
-void StacReader::validateSchema(NL::json stacJson)
+void StacReader::handleCatalog(NL::json stacJson, std::string catPath, bool isRoot)
 {
-    std::function<void(const nlohmann::json_uri&, nlohmann::json&)> fetch = schemaFetch;
-    nlohmann::json_schema::json_validator val(
-        fetch,
-        [](const std::string &, const std::string &) {}
-    );
-    if (!stacJson.contains("type"))
-        throw pdal_error("Invalid STAC json");
-    std::string type = stacJson.at("type").get<std::string>();
-    std::string schemaUrl;
-
-    if (type == "Feature")
+    stac::Catalog catalog(stacJson, catPath, *m_p->m_connector, *m_p->m_pool,
+        m_args->validateSchema);
+
+    if (catalog.init(*m_p->m_catFilters, m_args->rawReaderArgs,
+        m_args->schemaUrls, true))
     {
-        schemaUrl = m_args->featureSchemaUrl;
-        for (auto& extSchemaUrl: stacJson.at("stac_extensions"))
+        for (stac::Item& item: catalog.items())
+            addItem(item);
+    }
+    stac::ErrorList errors = catalog.errors();
+    if (errors.size())
```

Comment:
Looks like this is a "soft" sort of failure since we aren't `throw`ing in response to errors here.  A comment about why this failure is not fatal would help.